action/protocol/staking/protocol_test.go

```diff
@@ -0,0 +1,116 @@
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package staking
+
+import (
+	"math/big"
+	"testing"
+
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+
+	"github.com/iotexproject/iotex-core/test/identityset"
+)
+
+func TestDelegates(t *testing.T) {
+	require := require.New(t)
+
+	stk := NewProtocol()
+	require.NotNil(stk)
+
+	tests := []struct {
+		d          *Delegate
+		add, final *big.Int
+	}{
+		// invalid address
+		{
+			&Delegate{
+				Owner:   identityset.Address(2),
+				Address: "io1066kus4vlyvk0ljql39fzwqw0k22h7j8wmef2n",
+				Votes:   big.NewInt(100),
+				CanName: ToCandName([]byte("test1")),
+			},
+			big.NewInt(11), big.NewInt(111),
+		},
+		// negative votes
+		{
+			&Delegate{
+				Owner:         identityset.Address(2),
+				Address:       "io1066kus4vlyvk0ljql39fzwqw0k22h7j8wmef3n",
+				RewardAddress: "io1066kus4vlyvk0ljql39fzwqw0k22h7j8wmef3n",
+				Votes:         big.NewInt(-1),
+				CanName:       ToCandName([]byte("test1")),
+			},
+			big.NewInt(11), big.NewInt(111),
+		},
+		{
+			&Delegate{
+				Owner:         identityset.Address(1),
+				Address:       "io1066kus4vlyvk0ljql39fzwqw0k22h7j8wmef3n",
+				RewardAddress: "io1066kus4vlyvk0ljql39fzwqw0k22h7j8wmef3n",
+				Votes:         big.NewInt(100),
+				CanName:       ToCandName([]byte("test1")),
+			},
+			big.NewInt(11), big.NewInt(111),
+		},
+		{
+			&Delegate{
+				Owner:         identityset.Address(2),
+				Address:       "io1757z4d53408usrx2nf2vr5jh0mc5f5qm8nkre2",
+				RewardAddress: "io1757z4d53408usrx2nf2vr5jh0mc5f5qm8nkre2",
+				Votes:         big.NewInt(200),
+				CanName:       ToCandName([]byte("test2")),
+			},
+			big.NewInt(22), big.NewInt(222),
+		},
+		{
+			&Delegate{
+				Owner:         identityset.Address(3),
+				Address:       "io1mflp9m6hcgm2qcghchsdqj3z3eccrnekx9p0ms",
+				RewardAddress: "io1mflp9m6hcgm2qcghchsdqj3z3eccrnekx9p0ms",
+				Votes:         big.NewInt(300),
+				CanName:       ToCandName([]byte("test3")),
+			},
+			big.NewInt(33), big.NewInt(333),
+		},
+	}
+
+	require.Error(stk.addDelegate(tests[0].d))
+	require.Equal(ErrInvalidAmount, errors.Cause(stk.addDelegate(tests[1].d)))
+	tests = tests[2:]
+
+	// test add
+	for i, v := range tests {
+		require.NoError(stk.addDelegate(v.d))
+		require.Equal(i+1, len(stk.delegates))
+		require.True(stk.hasDelegate(v.d.Owner))
+		require.NoError(stk.addVote(v.d.Owner, v.add))
+		require.Equal(v.final, stk.delegates[v.d.Owner.String()].Votes)
+	}
+
+	// test delegate that does not exist
+	noName, err := address.FromBytes(make([]byte, 20))
+	require.NoError(err)
+	require.False(stk.hasDelegate(noName))
+	require.Equal(ErrInvalidAmount, errors.Cause(stk.addVote(tests[0].d.Owner, big.NewInt(-1))))
+	exceed := big.NewInt(0).Add(tests[0].final, big.NewInt(1))
```

Comment:
it is hard to follow tests[0] which has been modified in line 85

Reply:
ok, line 85 is to get rid of the first 2 fail case, so the rest of test only use the remaining 3 valid case